### ISSUE TITLE
Design

### ISSUE THREAD

**Issue by KomachiSion** [avalon566/sharding-scaling#33]

**Comment by KomachiSion** [comment]
My design is
![ShardingScalingLayer](https://raw.githubusercontent.com/KomachiSion/sharding-scaling/design/static/ShardingScalingLayer.png)

![ControllerProcess](https://raw.githubusercontent.com/KomachiSion/sharding-scaling/design/static/ControllerProcess.png)

**Comment by ssxlulu** [comment]
Hi, @KomachiSion ,I read the design carefully, and there are my suggestions:
1. I think data migration must pay attention to two points, one is the speed, and the other is the data  consistency.
2. Distributed system architecture may speed up data migration, but it makes it more difficult to ensure the data consistency.Sometimes,we pay more attention to the data consistency rather than speed.
3. In isomorphic system, like migrating data from MySQL to MySQL, I believe no tools can faster than the MySQL official tools like mysqldump or load data.So we use mysqldump in my company to migrate data from MySQL to MySQL. But we also need the migration tools like shardingscaling to migrate realtime data and work with the system to switch data sources.
4. Finally, we may refer to other products like [datax](https://github.com/alibaba/DataX/blob/master/introduction.md) of alibaba, for data migration, we may have something in common.

In conclusion, we may pay more attention to the data consistency when design.

**Comment by KomachiSion** [comment]
@ssxlulu 

> Distributed system architecture may speed up data migration, but it makes it more difficult to ensure the data consistency.Sometimes,we pay more attention to the data consistency rather than speed.

The master and worker can run in same process, which means they can run without distributed system. 

> In isomorphic system, like migrating data from MySQL to MySQL, I believe no tools can faster than the MySQL official tools like mysqldump or load data.So we use mysqldump in my company to migrate data from MySQL to MySQL. But we also need the migration tools like shardingscaling to migrate realtime data and work with the system to switch data sources.

It's a custom feature for yours, but it's a nice point for scaling component to make the migrate step to be configurable, users can choose which step they need. But it will be consider in future.

> Finally, we may refer to other products like datax of alibaba, for data migration, we may have something in common.

The component is designed for auto-scaling ShardingSphere finally, static migrate data is first and simple version, there will be many different feature.

At last, If you have better design or advise, we are welcome for your contributions.

**Comment by ssxlulu** [comment]
@KomachiSion ,thank you for your reply, and I believe it will be a greate design :smile:
